### ISSUE TITLE
torch.diag() fails on BoolTensors

### ISSUE THREAD

**Issue by Zalgo2462** [pytorch/pytorch#30923]
## 🐛 Bug

You cannot create a diagonal matrix from a boolean tensor in pytorch. 
## To Reproduce

1. Open a python terminal
1. `import torch`
1. `labels = torch.tensor([True, False, True], dtype=torch.bool)`
1. `diag = labels.diag()`

The error returned is `RuntimeError: _th_diag not supported on CPUType for Bool`.

If the tensor is a cuda tensor, then the error is `RuntimeError: _th_diag not supported on GPUType for Bool`

## Expected behavior

I expected `diag` to be hold the same value as
```
torch.tensor([
    [True, False, False],
    [False, False, False],
    [False, False, True]
], dtype=torch.bool)
````

## Environment
```
PyTorch version: 1.3.1
Is debug build: No
CUDA used to build PyTorch: 10.1.243

OS: Ubuntu 18.04.3 LTS
GCC version: (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
CMake version: version 3.10.2

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: GeForce RTX 2070
Nvidia driver version: 430.50
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.4

Versions of relevant libraries:
[pip3] numpy==1.17.2
[pip3] torch==1.3.1
[pip3] torchvision==0.4.2
[conda] Could not collect
```

## Additional context

cc @izdeby

**Comment by Zalgo2462** [comment]
For now, I am using uint8 tensors to get around this, but I run into the deprecation warning about indexing using uint8 tensors.

**Comment by zhangguanheng66** [comment]
@izdeby Any ideas?

**Comment by izdeby** [comment]
> @izdeby Any ideas?

`_th_diag` needs bool support. i've assigned this issue to myself, will fix.

**Comment by amirzaei1367** [comment]
a = torch.eye(3)
a = a>0

**Comment by H-Huang** [comment]
Fixed in #47455
